packages/api/src/controllers/generate.ts

```diff
@@ -1,33 +1,161 @@
+import crypto from "crypto";
+import { Request, RequestHandler, Router } from "express";
 import FormData from "form-data";
 import multer from "multer";
-import { BodyInit } from "node-fetch";
+import { BodyInit, Response } from "node-fetch";
+import promclient from "prom-client";
+import { v4 as uuid } from "uuid";
 import logger from "../logger";
 import { authorizer, validateFormData, validatePost } from "../middleware";
+import { AiGenerateLog, TextToImagePayload } from "../schema/types";
+import { db } from "../store";
+import { BadRequestError } from "../store/errors";
 import { fetchWithTimeout } from "../util";
 import { experimentSubjectsOnly } from "./experiment";
 import { pathJoin2 } from "./helpers";
 
 const AI_GATEWAY_TIMEOUT = 10 * 60 * 1000; // 10 minutes
 
+type AiGenerateType = AiGenerateLog["type"];
+
 const multipart = multer({
   storage: multer.memoryStorage(),
   limits: { fileSize: 10485760 }, // 10MiB
 });
 
+const aiGenerateDurationMetric = new promclient.Histogram({
+  name: "livepeer_api_ai_generate_duration",
+  buckets: [100, 500, 1000, 2500, 5000, 10000, 30000, 60000],
+  help: "Duration in milliseconds of the AI generation request",
+  labelNames: ["type", "complexity"], // see aiGenerateMetricComplexity
+});
+
+// We divide observations in "complexity" buckets. The complexity is only relevant on image generation pipelines that
+// have num images and/or inference steps params. The complexity is based on the product of them.
+function aiGenerateMetricComplexity(request: AiGenerateLog["request"]) {
+  const numInferenceSteps =
+    (request as TextToImagePayload)?.num_inference_steps ?? 50;
+  const numImages = (request as TextToImagePayload)?.num_images_per_prompt ?? 1;
+  // We divide by 200 to reduce cardinality of the metric.
+  // The range goes from 0 to `200 * 20 / 200` = 20 maximum cardinality
+  return Math.round((numInferenceSteps * numImages) / 200);
+}
+
 const app = Router();
 
 app.use(experimentSubjectsOnly("ai-generate"));
 
+function createPayload(
+  req: Request,
+  isJSONReq: boolean,
+  defaultModel: string,
+): [BodyInit, AiGenerateLog["request"]] {
+  const payload = {
+    model_id: defaultModel,
+    ...req.body,
+  };
+  if (isJSONReq) {
+    return [JSON.stringify(payload), payload];
+  }
+  if (!Array.isArray(req.files)) {
+    throw new BadRequestError("Expected an array of files");
+  }
+
+  const form = new FormData();
+  if (!("model_id" in req.body)) {
+    form.append("model_id", defaultModel);
+  }
+  for (const [key, value] of Object.entries(req.body)) {
+    form.append(key, value);
+  }
+  for (const file of req.files) {
+    form.append(file.fieldname, file.buffer, {
+      filename: file.originalname,
+      contentType: file.mimetype,
+      knownLength: file.size,
+    });
+    // We save only the hash of the file on the `payload` since that will be saved in the DB.
+    // TODO: Make this save the contents of the file in some OS and only reference them here.
```

Comment:
More to debug/reproduce the AI generations if we get weird reports for example.

The log of the input/outputs could also be useful to train/fine tune other AI models (e.g. if users gave feedback to those generations). So if we ever get to that point could want to do this. Could even consider to start storing them early so we start collecting data in case we want to do that later.